src/webserver/WsBase.h

```diff
@@ -322,26 +322,24 @@ class WsSession : public WsBase,
             return sendError(RPC::Error::rpcINTERNAL);
         }
 
-        std::string responseStr = boost::json::serialize(response);
         boost::json::array warnings;
-        auto warningFlag = false;
+        warnings.emplace_back(
+            "This is a clio server. clio only serves validated data. If you "
+            "want to talk to rippled, include 'ledger_index':'current' in your "
+            "request");
 
-        if (!dosGuard_.add(*ip, responseStr.size()))
-        {
-            warnings.emplace_back("Too many requests");
-            warningFlag = true;
-        }
         auto lastPublishAge = etl_->lastPublishAgeSeconds();
         if (lastPublishAge >= 60)
-        {
             warnings.emplace_back("This server may be out of date");
-            warningFlag = true;
-        }
-        // reserialize if a warning was appended
-        if (warningFlag)
+
+        auto& result = response["result"].as_object();
+        result["warnings"] = warnings;
+        std::string responseStr = boost::json::serialize(response);
+        if (!dosGuard_.add(*ip, responseStr.size()))
         {
-            auto& result = response["result"].as_object();
-            result["warning"] = warnings;
+            warnings.emplace_back("Too many requests");
+            result["warnings"] = warnings;
```

Comment:
This will only actually add warnings to result if `!dosGuard.add()`

I would restructure to something like 

```cpp
result["warnings"] = boost::json::array{};
boost::json::array& warnings = result["warnings"].as_array();
```

Reply:
Aren't they added on line 336?

Reply:
does ```result["warnings"] = warnings;``` cause a copy of the warnings array? I am now starting to think that it may be better to modify a reference to the response json directly.

Reply:
I'm concerned about the same thing @ethanlabelle 

@cjcobb23 have you tested this?

Reply:
So I'm actually pretty sure this won't get copied, since `boost::json::object::operator[]` returns a `value&`.

Reply:
But it also depends on how the internals of `operator[]` handle this operation.

Reply:
I tested this locally, and it works as expected.

Reply:
I tested it, everything seems to check out